### PR TITLE
Don't run scheduled tasks on shutdown, #16495

### PR DESCRIPTION
**Pull request by patriknw**
Behavior change in `Scheduler.close` that would be nice to do in 2.6.

Refs #16495

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/Scheduler.scala [hunk 1]
```diff
@@ -543,3 +549,14 @@ object Cancellable {
...
+
+object Scheduler {
+
+  /**
+   * If a `TaskRunOnClose` is used in `scheduleOnce` it will be run when the `Scheduler` is
+   * closed (`ActorSystem` shutdown). This is needed for the internal shutdown of dispatchers
+   * in Akka and is not intended to be used by end user applications, but it's public because
+   * a custom implementation of `Scheduler` must also implement this.
+   */
+  trait TaskRunOnClose extends Runnable
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/actor/Scheduler.scala]: `@NotForUserExtension` then?

**Review by johanandren — APPROVED** [review]

File: akka-actor-tests/src/test/scala/akka/actor/SchedulerSpec.scala [hunk 2]
```diff
@@ -483,10 +485,13 @@ class LightArrayRevolverSchedulerSpec extends AkkaSpec(SchedulerSpec.testConfRev
...
           val N = 1000000
-          (1 to N).foreach(_ => sched.scheduleOnce(delay)(counter.incrementAndGet()))
+          (1 to N).foreach(_ =>
+            sched.scheduleOnce(delay, new Scheduler.TaskRunOnClose {
+              override def run(): Unit = counter.incrementAndGet()
+            }))
           sched.close()
           Await.result(terminated, 3.seconds.dilated) should be > 10
-          awaitCond(counter.get == N)
+          awaitAssert(counter.get should ===(N))
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/actor/SchedulerSpec.scala]: I guess for the original "survive vicious enqueueing" check we could introduce a check that schedules 1000000 'regular' tasks and checks they all trigger within some deadline - but OK

File: akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala [hunk 3]
```diff
@@ -147,12 +147,27 @@ class LightArrayRevolverScheduler(config: Config, log: LoggingAdapter, threadFac
...
+    def runTask(task: Runnable): Unit = {
+      try task.run()
+      catch {
+        case e: InterruptedException => throw e
+        case _: SchedulerException   => // ignore terminated actors
+        case NonFatal(e)             => log.error(e, "exception while executing timer task")
+      }
+    }
+
+    Await.result(stop(), getShutdownTimeout).foreach {
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala]: first stops and then runs remaining tasks :+1:

**Review by jrudolph — APPROVED** [review]
LGTM, didn't know about this and I can imagine that this can lead to surprising behavior.

File: akka-actor/src/main/scala/akka/actor/Scheduler.scala (hunk 1 above)
Comment (raboof) [reply akka-actor/src/main/scala/akka/actor/Scheduler.scala]: As we allow custom schedulers that might be too strong..

**Review by raboof — APPROVED** [review]
Are there use cases for having users implement their own Scheduler? If not perhaps we should deprecate that, and `TaskRunOnClose` can be an implementation detail entirely (since we recommend users to use `system.registerOnTermination` or coordinated shutdown anyway)

**Pull request merged**